New module: WURFL device enrichment

andreacastello opened the pull request:
### 🔧 Type of changes
- [ ] new bid adapter
- [ ] bid adapter update
- [ ] new feature
- [ ] new analytics adapter
- [X] new module
- [ ] module update
- [ ] bugfix
- [ ] documentation
- [ ] configuration
- [ ] dependency update
- [ ] tech debt (test coverage, refactorings, etc.)

### ✨ What's the context?
The module populates missing or empty fields in **ortb2.device** with WURFL device detection data

### 🧠 Rationale behind the change

he WURFL module populates missing or empty fields in ortb2.device with the following data:
 - **make**: Manufacturer of the device (e.g., "Apple", "Samsung").
 - **model**: Device model (e.g., "iPhone 14", "Galaxy S22").
 - **os**: Operating system (e.g., "iOS", "Android").
 - **osv**: Operating system version (e.g., "16.0", "12.0").
 - **h**: Screen height in pixels.
 - **w**: Screen width in pixels.
 - **ppi**: Screen pixels per inch (PPI).
 - **pixelratio**: Screen pixel density ratio.
 - **devicetype**: Device type (e.g., mobile, tablet, desktop).

 - **Note**: If these fields are already populated in the bid request, the module will not overwrite them.

- Publisher-specific  enrichment
Device enrichment is selectively enabled for publishers based on their account ID. 
The module identifies publishers through the `getAccount()` method in the `AuctionContext` class.

### 🧪 Test plan
How do you know the changes are safe to ship to production?

### 🏎 Quality check
- [X] Are your changes following [our code style guidelines](https://github.com/prebid/prebid-server-java/blob/master/docs/developers/code-style.md)?
- [X] Are there any breaking changes in your code?
- [ ] Does your test coverage exceed 90%?
- [ ] Are there any erroneous console logs, debuggers or leftover code in your changes?

**Licensing Requirement for the WURFL Device Enrichment Module**
To use the *WURFL Device Enrichment Module* for Prebid Server, you must have a valid license for the *ScientiaMobile WURFL OnSite Java API*. This API is a core dependency of the module and is responsible for performing device detection and enrichment operations by leveraging ScientiaMobile's comprehensive WURFL device database.
You can obtain the required license from ScientiaMobile’s official product page:  [WURFL OnSite API for Java](https://www.scientiamobile.com/secondary-products/wurfl-onsite-api-for-java/)

### Maintainer contact

prebid@scientiamobile.com

Net-burst commented:
Hello, @andreacastello . We discussed this PR in the PBS committee meeting and this the summary:
- we can't merge the code which is not proven to be building and working;
- we can't merge the code, that won't be an active part of a build/test lifecycle and can unpredictably and silently break as a result;

To unblock this PR, you will need to have a way to make this code compilable and unit testable. I see several possible options:
- make a substitute "mock" library that will expose all of the API in the same packages so that the code can be built and unit tested; (I personally don't like this approach, but it was voiced in the PBS committee meeting)
- split the library into a public and private counterpart so that all of the public API is exposed through the public library, but unless the private library is in the classpath, that API will just noop; (I find this approach much better and much more maintainable)
- in any case, you will also need to provide documentation for the end users about public PBS builds not having this module in working state (or even it being omitted from `bundle` entirely), how to abb/substitute the private library to the project and how to build the project with this module included.

Until this is done, we'll be leaving this project in `do-not-merge` state. Feel free to convert thus PR to `draft` for now.

@bretg @osulzhenko @AntoxaAntoxic @CTMBNara FYI

Net-burst commented:
PBS-Go PR for reference: https://github.com/prebid/prebid-server/pull/4158

andreacastello commented:
Hello @Net-burst I've updated the PR so that the module can compile, run  the module's unit tests and execute when there's not a dependency to the licensed WURFL Java API. In this case the module calls an internal API mock (packages `com.scientiamobile.*` that provides sample data to the enrichment module.

I've also modified the module REAME to explain how to switch from this demo version to a licensed WURFL API usage.

Net-burst commented:
Thanks, @andreacastello! We'll take a look as soon as we have free cycles.

Net-burst commented on extra/modules/WURFL-devicedetection/README.md:
````diff
@@ -0,0 +1,268 @@
+```
+If the WURFL API dependency is not added, the module will compile a demo version that returns sample data, allowing basic testing without an WURFL Onsite Java API license.
+
+4 - Delete the `com` directory inside `extra/modules/WURFL-devicedetection/src/main/java`, which contains the WURFL Java API 
````
@andreacastello could you please move this code bundle into a publicly available library so that the end users will only need to change pom.xml dependency without the need to go into the code and risk breaking the project?

You can call that stub library `wurlf-demo` and add it as a pom.xml dependency. Then instruct users of the full version to just change the dependency from `wurfl-demo` to `wurfl`.

If you don't have a Maven repo set up, you can push it into Maven Central or use a GitHub Packages.

andreacastello replied:
@Net-burst I made the changes you requested. Now the mock is a publicly available dependency (from a Scientiamobile public repo)

Net-burst commented on extra/modules/pom.xml:
```diff
@@ -24,6 +24,7 @@
         <module>pb-response-correction</module>
         <module>greenbids-real-time-data</module>
         <module>pb-request-correction</module>
+        <module>WURFL-devicedetection</module>
```
Could you please use lowercase for your module name to be in line with other modules?

andreacastello replied:
Done

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfiguration.java:
```diff
@@ -0,0 +1,102 @@
+    private String createDownloadPath(WURFLDeviceDetectionConfigProperties configProperties) {
+        final String basePath = configProperties.getWurflFileDirPath();
+        final String fileName = configProperties.getWurflSnapshotUrl().endsWith(".xml.gz")
+                ? "new_wurfl.xml.gz"
+                : "new_wurfl.zip";
```
I don't like that hardcoded names.
Consider using `FileSyncerProperties` to configure WURFL updates, then truncate the file path if you only need the dir path.

andreacastello replied:
removed harcoded names

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfiguration.java:
```diff
@@ -0,0 +1,102 @@
+    private String createTempPath(WURFLDeviceDetectionConfigProperties configProperties) {
+        final String basePath = configProperties.getWurflFileDirPath();
+        final String fileName = configProperties.getWurflSnapshotUrl().endsWith(".xml.gz")
+                ? "temp_wurfl.xml.gz"
+                : "temp_wurfl.zip";
```
same

andreacastello replied:
removed this names too

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/AuctionRequestHeadersContext.java:
```diff
@@ -0,0 +1,31 @@
+        this.headers = headers;
+    }
+
+    public static AuctionRequestHeadersContext from(final CaseInsensitiveMultiMap headers) {
```
Remove `final` modifier here and other places. We don't use it for method parameters, it's assumed that nobody change the reference in method body

andreacastello replied:
removed all final modifiers in method parameters

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigProperties.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+import java.util.Set;
+
+@ConfigurationProperties(prefix = "hooks.modules." + WURFLDeviceDetectionModule.CODE)
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigProperties.java:
```diff
@@ -0,0 +1,61 @@
+            "is_phone",
+            "form_factor",
+            "pixel_density"
+    );
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigProperties.java:
```diff
@@ -0,0 +1,61 @@
+
+    boolean extCaps;
+
+    boolean wurflRunUpdater = true;
```
removed wurfl prefix from both variable names and config properties in yaml file

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigProperties.java:
```diff
@@ -0,0 +1,61 @@
+            "is_phone",
+            "form_factor",
+            "pixel_density"
+    );
```
removed

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigProperties.java:
```diff
@@ -0,0 +1,61 @@
+            "resolution_height",
+            "resolution_width",
+            "physical_form_factor"
+    );
```
done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigProperties.java:
```diff
@@ -0,0 +1,61 @@
+
+    boolean wurflRunUpdater = true;
+
+    List<String> allowedPublisherIds = List.of();
```
done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfiguration.java:
```diff
@@ -0,0 +1,102 @@
+@EnableConfigurationProperties(WURFLDeviceDetectionConfigProperties.class)
+public class WURFLDeviceDetectionConfiguration {
+
+    private static final Long DAILY_SYNC_INTERVAL = 86400000L;
```
created a new configurable property `update-frequency-in-hours`  in yaml for file syncer configuration

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfiguration.java:
```diff
@@ -0,0 +1,102 @@
+            final FileSyncer fileSyncer = createFileSyncer(configProperties, wurflService, vertx);
+            // Update process via file syncer starts with a delay because wurfl file has just been downloaded
+            vertx.setTimer(DAILY_SYNC_INTERVAL, ignored -> fileSyncer.sync());
+        }
```
Modified way the file syncer is initialized and used. Avoided use of initial file download using a wurfl engine method (syncer now takes care it this too).

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/AuctionRequestHeadersContext.java:
```diff
@@ -0,0 +1,31 @@
+
+    public static AuctionRequestHeadersContext from(final CaseInsensitiveMultiMap headers) {
+        final Map<String, String> headersMap = new HashMap<>();
+        if (Objects.isNull(headers)) {
```
done

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/exc/WURFLModuleConfigurationException.java:
`WURFLDeviceDetectionException`

andreacastello replied:
Done

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/HeadersResolver.java:
Don't forget to handle exception when calling this method or in the hook itself.
Also, is there a real need to escape this strings? If so, you might also need to escape `UserAgent` and `BrandVersion.version`

```
    public Map<String, String> resolve(Device device, Map<String, String> headers) {
        if (device == null && headers == null) {
            return Collections.emptyMap();
        }

        final Map<String, String> resolvedHeaders = resolveFromDevice(device);
        return MapUtils.isNotEmpty(resolvedHeaders)
                ? resolvedHeaders
                : headers;
    }

    private Map<String, String> resolveFromDevice(Device device) {
        if (device == null) {
            return Collections.emptyMap();
        }

        final Map<String, String> resolvedHeaders = new HashMap<>();
        if (device.getUa() != null) {
            resolvedHeaders.put(USER_AGENT, device.getUa());
        }
        resolvedHeaders.putAll(resolveFromSua(device.getSua()));

        return resolvedHeaders;
    }

    private Map<String, String> resolveFromSua(UserAgent sua) {
        if (sua == null) {
            return Collections.emptyMap();
        }

        final List<BrandVersion> brands = sua.getBrowsers();
        if (CollectionUtils.isEmpty(brands)) {
            return Collections.emptyMap();
        }

        final Map<String, String> headers = new HashMap<>();

        final String brandList = brandListAsString(brands);
        headers.put(SEC_CH_UA, brandList);
        headers.put(SEC_CH_UA_FULL_VERSION_LIST, brandList);

        final BrandVersion platform = sua.getPlatform();
        if (platform != null) {
            headers.put(SEC_CH_UA_PLATFORM, escape(platform.getBrand()));
            headers.put(SEC_CH_UA_PLATFORM_VERSION, escape(versionFromTokens(platform.getVersion())));
        }

        final String model = sua.getModel();
        if (StringUtils.isNotEmpty(model)) {
            headers.put(SEC_CH_UA_MODEL, escape(model));
        }

        final String arch = sua.getArchitecture();
        if (StringUtils.isNotEmpty(arch)) {
            headers.put(SEC_CH_UA_ARCH, escape(arch));
        }

        final Integer mobile = sua.getMobile();
        if (mobile != null) {
            headers.put(SEC_CH_UA_MOBILE, "?" + mobile);
        }

        return headers;
    }

    private String brandListAsString(List<BrandVersion> versions) {
        return versions.stream()
                .filter(brandVersion -> brandVersion.getBrand() != null)
                .map(brandVersion -> "%s;v=\"%s\"".formatted(
                        escape(brandVersion.getBrand()),
                        versionFromTokens(brandVersion.getVersion())))
                .collect(Collectors.joining(", "));
    }

    private static String escape(String value) {
        try {
            return '"' + StringEscapeUtils.escapeJava(value) + '"';
        } catch (Exception e) {
            throw new WURFLDeviceDetectionException(e);
        }
    }

    private static String versionFromTokens(List<String> tokens) {
        if (CollectionUtils.isEmpty(tokens)) {
            return StringUtils.EMPTY;
        }

        return tokens.stream()
                .filter(StringUtils::isNotEmpty)
                .collect(Collectors.joining("."));
    }
```

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/PlatformNameVersion.java:
No need for this class

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/AccountValidator.java:
No need for this class

andreacastello replied:
Removed it

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/ExtWURFLMapper.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.Objects;
+import java.util.Set;
+
+@Builder
+@Slf4j
```
Same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/ExtWURFLMapper.java:
```diff
@@ -0,0 +1,73 @@
+    private static final String WURFL_ID_PROPERTY = "wurfl_id";
+
+    public JsonNode mapExtProperties() {
+
+        final ObjectMapper objectMapper = new ObjectMapper();
```
Create it once or use `ObjectMapperProvider.mapper()` (see if its configuration breaks your logic)

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/ExtWURFLMapper.java:
```diff
@@ -0,0 +1,73 @@
+
+            if (addExtCaps) {
+
+                staticCaps.stream()
+                        .map(sc -> {
+                            try {
+                                return Map.entry(sc, wurflDevice.getCapability(sc));
+                            } catch (Exception e) {
+                                log.error("Error getting capability for {}: {}", sc, e.getMessage());
+                                return Map.entry(sc, NULL_VALUE_TOKEN);
+                            }
+                        })
+                        .filter(entry -> entry.getValue() != null
+                                && !NULL_VALUE_TOKEN.equals(entry.getValue()))
+                        .forEach(entry -> wurflNode.put(entry.getKey(), entry.getValue()));
```
Use `for loop` instead of streamApi approach

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/ExtWURFLMapper.java:
```diff
@@ -0,0 +1,73 @@
+                                && !NULL_VALUE_TOKEN.equals(entry.getValue()))
+                        .forEach(entry -> wurflNode.put(entry.getKey(), entry.getValue()));
+
+                virtualCaps.stream()
+                        .map(vc -> {
+                            try {
+                                return Map.entry(vc, wurflDevice.getVirtualCapability(vc));
+                            } catch (Exception e) {
+
+                                log.warn("Could not fetch virtual capability {}", vc);
+                                return null;
+                            }
+                        })
+                        .filter(Objects::nonNull)
+                        .forEach(entry -> wurflNode.put(entry.getKey(), entry.getValue()));
```
Use `for loop` instead of streamApi approach

CTMBNara commented:
@andreacastello Please look at other classes and methods and try to simplify and remove redundancy in them like in `AccountValidator` or `PlatformNameVersion`.

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfiguration.java:
```diff
@@ -0,0 +1,96 @@
+
+    private String createTempPath(WURFLDeviceDetectionConfigProperties configProperties) {
+        final String basePath = configProperties.getFileDirPath();
+        String fileName = WURFLEngineInitializer.extractWURFLFileName(configProperties.getFileSnapshotUrl());
+        fileName = "tmp_" + fileName;
```
> ```
>          final String fileName = "tmp_" + WURFLEngineInitializer.extractWURFLFileName(configProperties.getFileSnapshotUrl());
> ```

This makes checkstyle fail: using
``` 
final String fileName = "tmp_"
                + WURFLEngineInitializer.extractWURFLFileName(configProperties.getFileSnapshotUrl());
```

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/ExtWURFLMapper.java:
```diff
@@ -0,0 +1,73 @@
+            }
+        } catch (Exception e) {
+            log.error("Exception while updating EXT");
+        }
```
Completely reviewed that part.

andreacastello commented:
> @andreacastello Please look at other classes and methods and try to simplify and remove redundancy in them like in `AccountValidator` or `PlatformNameVersion`.

I've removed both classes in favor of internal methods usage

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/WURFLEngineInitializer.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Slf4j
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/WURFLEngineInitializer.java:
```diff
@@ -0,0 +1,106 @@
+            throw new WURFLModuleConfigurationException(failedCheckMessage);
+        }
+
+    }
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/WURFLEngineInitializer.java:
```diff
@@ -0,0 +1,106 @@
+
+@Slf4j
+@Builder
+public class WURFLEngineInitializer {
```
Done. See above.

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/HeadersResolver.java:
```diff
@@ -0,0 +1,131 @@
+    static final String SEC_CH_UA_ARCH = "Sec-CH-UA-Arch";
+    static final String SEC_CH_UA_MODEL = "Sec-CH-UA-Model";
+    static final String SEC_CH_UA_FULL_VERSION_LIST = "Sec-CH-UA-Full-Version-List";
+    static final String USER_AGENT = "User-Agent";
```
Moved headers to HttpUtil class

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/HeadersResolver.java:
```diff
@@ -0,0 +1,131 @@
+
+    private String brandListAsString(List<BrandVersion> versions) {
+
+        final String brandNameString = versions.stream()
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfiguration.java:
```diff
@@ -0,0 +1,95 @@
+                                        WURFLService wurflService, Vertx vertx) {
+        final FileSyncerProperties fileSyncerProperties = createFileSyncerProperties(configProperties);
+        return FileUtil.fileSyncerFor(wurflService, fileSyncerProperties, vertx);
+    }
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfiguration.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Bean
+    public WURFLDeviceDetectionModule wurflDeviceDetectionModule(WURFLDeviceDetectionConfigProperties
+                                                                         configProperties, Vertx vertx) {
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/WURFLEngineUtils.java:
```diff
@@ -0,0 +1,84 @@
+            "resolution_width",
+            "physical_form_factor");
+
+    public static WURFLEngine initializeEngine(WURFLDeviceDetectionConfigProperties configProperties,
+                                               String wurflInFilePath) {
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/OrtbDeviceUpdater.java:
```diff
@@ -0,0 +1,248 @@
+            try {
+                pxRatio = new BigDecimal(fromWurfl);
+                return UpdateResult.updated(pxRatio);
+            } catch (NullPointerException e) {
```
You're right, `fromWurfl` has been previously checked thus it's impossible to get `NPE`. Removed try/catch

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHook.java:
Refactoring applied. Some possible excpetions are handled inside the updater class

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHook.java:
```diff
@@ -0,0 +1,140 @@
+        this.ortbDeviceUpdater = new OrtbDeviceUpdater();
+        this.addExtCaps = configProperties.isExtCaps();
+        this.allowedPublisherIDs = configProperties.getAllowedPublisherIds().stream()
+                .collect(Collectors.toMap(item -> item, item -> item));
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHook.java:
```diff
@@ -0,0 +1,140 @@
+                final Device updatedDevice = ortbDeviceUpdater.update(ortbDevice, wurflDevice.get(),
+                        wurflService.getAllCapabilities(),
+                        wurflService.getAllVirtualCapabilities(),
+                        addExtCaps);
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/OrtbDeviceUpdater.java:
Refactoring applied

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/ExtWURFLMapper.java:
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionModule.java:
```diff
@@ -0,0 +1,29 @@
+
+    public WURFLDeviceDetectionModule(List<? extends Hook<?, ? extends InvocationContext>> hooks) {
+        this.hooks = hooks;
+
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/HeadersResolver.java:
```diff
@@ -0,0 +1,102 @@
+        return headers;
+    }
+
+    private String brandListAsString(List<BrandVersion> versions) {
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/HeadersResolver.java:
```diff
@@ -0,0 +1,102 @@
+                : headers;
+    }
+
+    private Map<String, String> resolveFromDevice(Device device) {
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/HeadersResolver.java:
```diff
@@ -0,0 +1,102 @@
+
+public class HeadersResolver {
+
+    public Map<String, String> resolve(Device device, Map<String, String> headers) {
```
Done

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/OrtbDeviceUpdater.java:
```diff
@@ -0,0 +1,235 @@
+            return densityAsString != null
+                ? new BigDecimal(densityAsString)
+                : null;
+        } catch (CapabilityNotDefinedException e) {
```
`NumberFormatException`

andreacastello replied:
Done (see comment on exception handling in the following remark

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/OrtbDeviceUpdater.java:
Why do you catch exceptions for some fields and not others?

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigPropertiesTest.java:
```diff
@@ -0,0 +1,55 @@
+class WURFLDeviceDetectionConfigPropertiesTest {
+
+    @Test
+    void shouldInitializeWithEmptyValues() {
```
add `public` modifier to all tests

andreacastello replied:
Done to all test classes and methods

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigPropertiesTest.java:
No need for this test class

andreacastello replied:
Removed

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigurationTest.java:
No need for this test class

andreacastello replied:
Removed

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/mock/WURFLDeviceMock.java:
Use `@Mock` with `given(...).willReturn(...)` instead of this class

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/AuctionRequestHeadersContextTest.java:
```diff
@@ -0,0 +1,65 @@
+class AuctionRequestHeadersContextTest {
+
+    @Test
+    void fromShouldHandleNullHeaders() {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/WURFLEngineUtilsTest.java:
```diff
@@ -0,0 +1,82 @@
+    private WURFLDeviceDetectionConfigProperties configProperties;
+
+    @Test
+    void extractWURFLFileNameShouldReturnCorrectFileName() {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfigPropertiesTest.java:
```diff
@@ -0,0 +1,55 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class WURFLDeviceDetectionConfigPropertiesTest {
```
`public class`

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/model/AuctionRequestHeadersContextTest.java:
```diff
@@ -0,0 +1,65 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class AuctionRequestHeadersContextTest {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/HeadersResolverTest.java:
```diff
@@ -0,0 +1,172 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class HeadersResolverTest {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/resolver/HeadersResolverTest.java:
```diff
@@ -0,0 +1,172 @@
+    private static final String TEST_USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36";
+
+    @Test
+    void resolveWithNullDeviceShouldReturnOriginalHeaders() {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/OrtbDeviceUpdaterTest.java:
```diff
@@ -0,0 +1,307 @@
+import static org.prebid.server.hooks.modules.com.scientiamobile.wurfl.devicedetection.mock.WURFLDeviceMock.WURFLDeviceMockFactory.mockUnknownDevice;
+
+@ExtendWith(MockitoExtension.class)
+class OrtbDeviceUpdaterTest {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionEntrypointHookTest.java:
```diff
@@ -0,0 +1,81 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class WURFLDeviceDetectionEntrypointHookTest {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionEntrypointHookTest.java:
```diff
@@ -0,0 +1,81 @@
+    private InvocationContext context;
+
+    @BeforeEach
+    void setUp() {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionEntrypointHookTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @BeforeEach
+    void setUp() {
+        payload = mock(EntrypointPayload.class);
+        context = mock(InvocationContext.class);
```
Use `@Mock` instead

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionModuleTest.java:
```diff
@@ -0,0 +1,40 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class WURFLDeviceDetectionModuleTest {
```
same

CTMBNara replied:
`public class WURFLDeviceDetectionModuleTest`

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionModuleTest.java:
```diff
@@ -0,0 +1,40 @@
+class WURFLDeviceDetectionModuleTest {
+
+    @Test
+    void codeShouldReturnCorrectModuleCode() {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHookTest.java:
```diff
@@ -0,0 +1,220 @@
+                .build();
+        final AuctionRequestHeadersContext headersContext = AuctionRequestHeadersContext.from(headers);
+
+        // when
```
it's still `given` step

andreacastello replied:
Removed wrong comment

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHookTest.java:
```diff
@@ -0,0 +1,220 @@
+                .build();
+        final AuctionRequestHeadersContext headersContext = AuctionRequestHeadersContext.from(headers);
+
+        // when
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHookTest.java:
```diff
@@ -0,0 +1,220 @@
+                .build();
+        final AuctionRequestHeadersContext headersContext = AuctionRequestHeadersContext.from(headers);
+
+        // when
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHookTest.java:
```diff
@@ -0,0 +1,220 @@
+    private WURFLDeviceDetectionRawAuctionRequestHook target;
+
+    @BeforeEach
+    void setUp() {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHookTest.java:
```diff
@@ -0,0 +1,220 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class WURFLDeviceDetectionRawAuctionRequestHookTest {
```
same

CTMBNara commented on extra/modules/wurfl-devicedetection/src/test/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+
+        final WURFLService spyWurflService = spy(wurflService);
+
+        // Mock the createEngine method to return our mock engine
```
No need for this comment

andreacastello replied:
Removed

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/config/WURFLDeviceDetectionConfiguration.java:
```diff
@@ -0,0 +1,100 @@
+                new WURFLDeviceDetectionEntrypointHook(),
+                new WURFLDeviceDetectionRawAuctionRequestHook(wurflService, configProperties,
+                        new JacksonMapper(ObjectMapperProvider.mapper()))));
+    }
```
Done

andreacastello commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHook.java:
```diff
@@ -0,0 +1,119 @@
+        if (bidRequest == null || bidRequest.getDevice() == null) {
+            logger.warn("BidRequest is null");
+            return noActionResult();
+        }
```
Done

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/WURFLDeviceDetectionRawAuctionRequestHook.java:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        final BidRequest bidRequest = auctionRequestPayload.bidRequest();
+        if (bidRequest.getDevice() == null) {
+            logger.warn("BidRequest is null");
+            return noActionResult();
+        }
+
+        final Device device = bidRequest.getDevice();
+        if (device == null) {
+            logger.warn("Device is null");
+            return noActionResult();
+        }
```
duplicate

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/OrtbDeviceUpdater.java:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    private Integer getWurflDeviceType() {
+
```
Remove empty line

CTMBNara commented on extra/modules/wurfl-devicedetection/src/main/java/org/prebid/server/hooks/modules/com/scientiamobile/wurfl/devicedetection/v1/OrtbDeviceUpdater.java:
```diff
@@ -0,0 +1,259 @@
+            if (wurflDevice.getCapabilityAsBool("is_ott")) {
+                return 7;
+            }
+        } catch (CapabilityNotDefinedException | VirtualCapabilityNotDefinedException | NumberFormatException e) {
+            logger.warn("Failed to determine device type from WURFL device capabilities", e);
+        }
+        return null;
```
```
try {
    ...
} catch (...) {
    ...
    return null;
}
```

osulzhenko commented:
@andreacastello could you please resolve conflicts?

andreacastello commented:
> @andreacastello could you please resolve conflicts?

@Net-burst I resolved the conflict and bumped the version of the WURFL module, it should be ok now
